arrow-libs/resilience/arrow-resilience/src/commonMain/kotlin/arrow/resilience/Schedule.kt

```diff
@@ -406,46 +409,86 @@ public value class Schedule<in Input, out Output>(public val step: ScheduleStep<
 }
 
 /**
- * Retries [action] using any [Throwable] that occurred as the input to the [Schedule].
+ * Retries [action] using any [E] that occurred as the input to the [Schedule].
  * It will throw the last exception if the [Schedule] is exhausted, and ignores the output of the [Schedule].
  */
-public suspend fun <A> Schedule<Throwable, *>.retry(action: suspend () -> A): A =
-  retryOrElse(action) { e, _ -> throw e }
+public suspend inline fun <reified E: Throw, Throw: Throwable, A> Schedule<Throw, *>.retry(
+  noinline action: suspend () -> A
+): A = retry(E::class, action)
 
 /**
- * Retries [action] using any [Throwable] that occurred as the input to the [Schedule].
+ * Retries [action] using any [E] that occurred as the input to the [Schedule].
+ * It will throw the last exception if the [Schedule] is exhausted, and ignores the output of the [Schedule].
+ */
+public suspend fun <E: Throw, Throw: Throwable, A> Schedule<Throw, *>.retry(
```

Comment:
I am not sure I understand the approach `<E: Throw, Throw: Throwable` used. I have the feeling I am overlooking something obvious 🤔


Besides that this looks great! Wanted to update my PR to use `KCLass<E>` as we discussed on Slack (don't think I pushed yet), but I'm not sure I understand the need for this indirection in the generic definition.

Not sure I understand why this isn't the equivalent of `<E :Throwable`? 🤔

Reply:
The problem is in the receiver type. Let's consider both "extremes":
- If we make the receiver `Schedule<Throwable, *>`, then we can only call `retry` with that particular schedule type. But we may want to make it possible to also call it with `Schedule<IllegalStateException, *>`, for example.
- If we make the receiver `Schedule<E, *>`, then the filtering is actually useless.

This was we can call the function with any schedule with an input which is a subtype of `Throwable`, and then further filter the cases for `E` which are a subset of those. The variant with `E: Throwable, Throw: Throwable` would also compile, maybe that's the way to go?

Reply:
Ah, okay got it. So having the indirection inside the generics, allows getting away with a single impl. That's a really awesome trick!!

To be honest, I think `<E: Throwable, Throw: Throwable>` might throw users off in Dokka, or looking at the source code. So I guess I prefer a little duplication, rather than a more complex signature from a user point of view.

This is super nitty though 😅 I like your solution better, but wondering if users would trip of seeing that signature. One of my goals, also with the Schedule refactor was to make Arrows internals simpler, and easy to understand. So I often also choose more explicit code style in Arrow when working towards 1.x.x. It's a tiny detail, but still I feel it gives quite some value.

WDYT? Curious to get lang designers opinion on that 😁